### PR TITLE
EtcdConfigurationSource

### PR DESCRIPTION
**Pull request by spoon16**
An Etcd watched configuration source very similar to the ZooKeeperConfigurationSource.

Uses the Boon Etcd client https://github.com/boonproject/boon/blob/master/etcd/README.md.  Seems to be the most widely adopted Etcd client.

Tests included.

### CODE REVIEW

**Comment by spoon16** [comment]
Build failed because of a JVM error... or maybe a bunch of failing unit tests unrelated to the Etcd module.

Unless I am missing something.

**Comment by howardyuan** [comment]
There're two pull-requests builds. One of them was not configured with
proper Gradle opts so it runs out Permgen. I'll see how to either disable
the second one or add proper GRADLE_OPTS.

File: archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        public void handle(Response updateResponse) {
+            if (updateResponse.wasError()) {
+                logger.error("Etcd failed with an error response: %s", updateResponse);
+            }
+
+            final Map<String, Object> create = Maps.newHashMap();
+            final Map<String, Object> set = Maps.newHashMap();
+            final Map<String, Object> delete = Maps.newHashMap();
+
+            final String action = updateResponse.action().toLowerCase();
```
Comment (howardyuan) [inline archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java]: Can the action() return null when it's error?

File: archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+    private final String configPath;
+
+    private Handler<Response> updateHandler = new Handler<Response>() {
+        @Override
+        public void handle(Response updateResponse) {
+            if (updateResponse.wasError()) {
+                logger.error("Etcd failed with an error response: %s", updateResponse);
+            }
+
+            final Map<String, Object> create = Maps.newHashMap();
```
Comment (howardyuan) [inline archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java]: I was a little puzzled by three maps here for a while. For sure, we will get two empty maps and one map with one entry in there? Not really sure if I have a better idea yet, just want to point out.

File: archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+            final Map<String, Object> delete = Maps.newHashMap();
+
+            final String action = updateResponse.action().toLowerCase();
+            final Node node = updateResponse.node();
+
+            if (node != null ) {
+                final String etcdKey = node.key();
+                final String sourceKey = Iterables.getLast(keySplitter.split(etcdKey));
+                final String value = node.getValue();
+                valueCache.put(sourceKey, value);
```
Comment (howardyuan) [inline archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java]: what if it's a delete, then should you remove the entry from the cache?

Comment (spoon16) [inline archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java]: good catch, will fix

File: archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java (hunk 2 above)
Comment (spoon16) [inline archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java]: Yeah, not sure what would be better here. I could do something like:

```java

```

File: archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java (hunk 1 above)
Comment (spoon16) [inline archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java]: only if I duplicate the etcd.waitRecursive call on l:89

It'd probably be best to capture the wasError condition, throw an exception and have a finally that calls etcd.waitRecursive.  But I'm not sure that really makes things more readable.  I'll just duplicate l:89 and return early.

File: archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java (hunk 3 above)
Comment (spoon16) [inline archaius-etcd/src/main/java/com/netflix/config/source/EtcdConfigurationSource.java]: fixed and added a test to validate

**Comment by spoon16** [comment]
@howardyuan, tried to address your feedback with the following changes.

- fixed bug where delete wouldn't remove the value from the `EtcdConfigurationSource`
- added a unit test to validate delete behavior
- refactored the `updateHandler` in `EtcdConfigurationSource`... tried to improve readability by not always instantiating three maps

**Comment by howardyuan** [comment]
Looks good to me. Will leave the PR open for a few days to see if anybody else have any more comments.

**Comment by chandrapersonal** [comment]
can i pull the changes.

**Comment by spoon16** [comment]
@howardyuan merge?

**Comment by chandrapersonal** [comment]
why there is no case for node action "compareandswap" in etcd UpdateHandler. we have node actions create,set and delete action on UpdateHandler . but no  compareandswap. Is there any reason for not have that action.

**Comment by spoon16** [comment]
@chandrapersonal I don't think "compareandswap" is a type of update.  It's part of the Etcd update API and the side effect is that an update notification is emitted, either set or create.

**Pull request merged**